File: Backend/src/modules/routes/routes.service.ts
import { getDirectionsRoute } from '../../services/googleRoutesDirectionsApi.service';

const GOOGLE_MAPS_API_KEY = process.env.GOOGLE_MAPS_API_KEY || '';

export interface Route {
  id: string;
  origin: Waypoint;
  destination: Waypoint;
  waypoints?: Waypoint[];
  distance: number; // em metros
  duration: number; // em segundos
  distanceText: string;
  durationText: string;
  polyline: string;
  steps: RouteStep[];
  travelMode: TravelModeType;
  bounds?: {
    northeast: { lat: number; lng: number };
    southwest: { lat: number; lng: number };
  };
  warnings?: string[];
  copyrights?: string;
}

export interface Waypoint {
  placeId?: string;
  name?: string;
  latitude: number;
  longitude: number;
}

export interface RouteStep {
  instruction: string;
  distance: number;
  distanceText: string;
  duration: number;
  durationText: string;
  startLocation: { lat: number; lng: number };
  endLocation: { lat: number; lng: number };
  travelMode: string;
  maneuver?: string;
}

export type TravelModeType = 'driving' | 'walking' | 'bicycling' | 'transit';

export interface DistanceMatrixResult {
  origin: Waypoint;
  destination: Waypoint;
  distance: number;
  distanceText: string;
  duration: number;
  durationText: string;
  status: string;
}



export class RoutesService {
  async calculateRoute(
    origin: Waypoint,
    destination: Waypoint,
    travelMode: TravelModeType = 'driving',
    waypoints?: Waypoint[],
    optimize?: boolean,
    language?: string
  ): Promise<Route | null> {
    try {
      // Nova API não suporta múltiplos waypoints ainda, só origem e destino
      const apiResponse = await getDirectionsRoute({
        origin: { lat: origin.latitude, lng: origin.longitude },
        destination: { lat: destination.latitude, lng: destination.longitude },
        travelMode,
        languageCode: 'pt-PT',
      });

      if (!apiResponse.routes || apiResponse.routes.length === 0) {
        return null;
      }
      const route = apiResponse.routes[0];
      const leg = route.legs && route.legs[0];

      // Calcular distância e duração totais
      const totalDistance = leg?.distanceMeters || 0;
      const totalDuration = leg?.duration ? (typeof leg.duration === 'string' ? this.parseDuration(leg.duration) : leg.duration) : 0;
      const allSteps: RouteStep[] = (leg?.steps || []).map((step: any) => ({
        instruction: step.navigationInstruction?.instructions || '',
        distance: step.distanceMeters || 0,
        distanceText: step.distanceMeters ? this.formatDistance(step.distanceMeters) : '',
        duration: step.duration ? (typeof step.duration === 'string' ? this.parseDuration(step.duration) : step.duration) : 0,
        durationText: step.duration ? this.formatDuration(typeof step.duration === 'string' ? this.parseDuration(step.duration) : step.duration) : '',
        startLocation: step.startLocation,
        endLocation: step.endLocation,
        travelMode: step.travelMode,
        maneuver: step.navigationInstruction?.maneuver,
      }));

      return {
        id: `route_${Date.now()}`,
        origin,
        destination,
        waypoints,
        distance: totalDistance,
        duration: totalDuration,
        distanceText: this.formatDistance(totalDistance),
        durationText: this.formatDuration(totalDuration),
        polyline: route.polyline?.encodedPolyline || '',
        steps: allSteps,
        travelMode,
        bounds: undefined, // Nova API não retorna bounds diretamente
        warnings: undefined,
        copyrights: undefined
      };
    } catch (error) {
      console.error('Erro ao calcular rota (Routes API):', error);
      return null;
    }
  }

  // getOptimizedRoute não suportado na nova API
  async getOptimizedRoute(): Promise<null> {
    throw new Error('getOptimizedRoute não suportado pela nova Google Routes API');
  }

  // getAlternativeRoutes não suportado na nova API
  async getAlternativeRoutes(): Promise<Route[]> {
    throw new Error('getAlternativeRoutes não suportado pela nova Google Routes API');
  }

  async getDistanceMatrix(
    origins: Waypoint[],
    destinations: Waypoint[],
    travelMode: TravelModeType = 'driving',
    language?: string
  ): Promise<DistanceMatrixResult[]> {
    try {
      const { getRouteMatrix } = await import('../../services/googleRoutesApi.service');
      const data = await getRouteMatrix(origins, destinations, travelMode);
      if (!Array.isArray(data)) {
        console.error('Resposta inesperada da Routes API:', data);
        return [];
      }
      const results: DistanceMatrixResult[] = data.map((item: any) => {
        // Considera status vazio/objeto como válido se houver distanceMeters
        let statusString = 'OK';
        if (typeof item.status === 'string') statusString = item.status;
        if (typeof item.status === 'object' && Object.keys(item.status).length > 0 && item.status.code) statusString = item.status.code;
        return {
          origin: origins[item.originIndex],
          destination: destinations[item.destinationIndex],
          distance: item.distanceMeters || 0,
          distanceText: item.distanceMeters ? this.formatDistance(item.distanceMeters) : 'N/A',
          duration: item.duration ? (typeof item.duration === 'string' ? this.parseDuration(item.duration) : item.duration.seconds) : 0,
          durationText: item.duration ? this.formatDuration(typeof item.duration === 'string' ? this.parseDuration(item.duration) : item.duration.seconds) : 'N/A',
          status: statusString,
        };
      });
      if (results.length === 0) {
        console.warn('⚠️ [getDistanceMatrix] Nenhum resultado retornado da nova Routes API');
      }
      return results;
    } catch (error: any) {
      console.error('Erro ao calcular matriz de distâncias (Routes API):', error?.response?.data || error);
      return [];
    }
  }

  // Utilitário para parsear string de duração tipo '3600s' para segundos
  private parseDuration(duration: string): number {
    if (!duration) return 0;
    const match = duration.match(/(\d+)s/);
    return match ? parseInt(match[1], 10) : 0;
  }

  // getTravelTimeWithTraffic não suportado na nova API
  async getTravelTimeWithTraffic(): Promise<null> {
    throw new Error('getTravelTimeWithTraffic não suportado pela nova Google Routes API');
  }

  decodePolyline(encoded: string): { lat: number; lng: number }[] {
    const points: { lat: number; lng: number }[] = [];
    let index = 0;
    let lat = 0;
    let lng = 0;

    while (index < encoded.length) {
      let shift = 0;
      let result = 0;
      let byte: number;

      do {
        byte = encoded.charCodeAt(index++) - 63;
        result |= (byte & 0x1f) << shift;
        shift += 5;
      } while (byte >= 0x20);

      const dlat = result & 1 ? ~(result >> 1) : result >> 1;
      lat += dlat;

      shift = 0;
      result = 0;

      do {
        byte = encoded.charCodeAt(index++) - 63;
        result |= (byte & 0x1f) << shift;
        shift += 5;
      } while (byte >= 0x20);

      const dlng = result & 1 ? ~(result >> 1) : result >> 1;
      lng += dlng;

      points.push({
        lat: lat / 1e5,
        lng: lng / 1e5
      });
    }

    return points;
  }

  private formatDistance(meters: number): string {
    if (meters >= 1000) {
      return `${(meters / 1000).toFixed(1)} km`;
    }
    return `${meters} m`;
  }

  private formatDuration(seconds: number): string {
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);

    if (hours > 0) {
      return `${hours} h ${minutes} min`;
    }
    return `${minutes} min`;
  }

  /**
   * Get simple distance between two points (for itinerary items)
   */
  async getDistance(
    origin: { latitude: number; longitude: number },
    destination: { latitude: number; longitude: number },
    travelMode: TravelModeType = 'driving'
  ): Promise<{ distance: number; distanceText: string; duration: number; durationText: string } | null> {
    try {
      const originWaypoint: Waypoint = { latitude: origin.latitude, longitude: origin.longitude };
      const destWaypoint: Waypoint = { latitude: destination.latitude, longitude: destination.longitude };
      
      const results = await this.getDistanceMatrix([originWaypoint], [destWaypoint], travelMode);
      
      if (results.length > 0 && results[0].distance > 0) {
        return {
          distance: results[0].distance,
          distanceText: results[0].distanceText,
          duration: results[0].duration,
          durationText: results[0].durationText
        };
      }
      
      return null;
    } catch (error) {
      console.error('Error calculating distance:', error);
      return null;
    }
  }
}

export const routesService = new RoutesService();

File: ui/lib/services/routes_service.dart
import 'api_service.dart';

class Waypoint {
  final String? placeId;
  final String? name;
  final double latitude;
  final double longitude;

  Waypoint({
    this.placeId,
    this.name,
    required this.latitude,
    required this.longitude,
  });

  Map<String, dynamic> toJson() => {
    if (placeId != null) 'placeId': placeId,
    if (name != null) 'name': name,
    'latitude': latitude,
    'longitude': longitude,
  };
}

class TravelInfo {
  final int durationMinutes;
  final String durationText;
  final int distanceMeters;
  final String distanceText;
  final String travelMode;

  TravelInfo({
    required this.durationMinutes,
    required this.durationText,
    required this.distanceMeters,
    required this.distanceText,
    required this.travelMode,
  });

  factory TravelInfo.fromJson(Map<String, dynamic> json) {
    return TravelInfo(
      durationMinutes: ((json['duration'] ?? 0) / 60).round(),
      durationText: json['durationText'] ?? '',
      distanceMeters: json['distance'] ?? 0,
      distanceText: json['distanceText'] ?? '',
      travelMode: json['travelMode'] ?? 'driving',
    );
  }
}

class RouteInfo {
  final String id;
  final Waypoint origin;
  final Waypoint destination;
  final int distance;
  final int duration;
  final String distanceText;
  final String durationText;
  final String polyline;
  final String travelMode;

  RouteInfo({
    required this.id,
    required this.origin,
    required this.destination,
    required this.distance,
    required this.duration,
    required this.distanceText,
    required this.durationText,
    required this.polyline,
    required this.travelMode,
  });

  factory RouteInfo.fromJson(Map<String, dynamic> json) {
    return RouteInfo(
      id: json['id'] ?? '',
      origin: Waypoint(
        latitude: json['origin']['latitude'] ?? 0.0,
        longitude: json['origin']['longitude'] ?? 0.0,
      ),
      destination: Waypoint(
        latitude: json['destination']['latitude'] ?? 0.0,
        longitude: json['destination']['longitude'] ?? 0.0,
      ),
      distance: json['distance'] ?? 0,
      duration: json['duration'] ?? 0,
      distanceText: json['distanceText'] ?? '',
      durationText: json['durationText'] ?? '',
      polyline: json['polyline'] ?? '',
      travelMode: json['travelMode'] ?? 'driving',
    );
  }
}

class DistanceMatrixResult {
  final int distance;
  final String distanceText;
  final int duration;
  final String durationText;
  final String status;

  DistanceMatrixResult({
    required this.distance,
    required this.distanceText,
    required this.duration,
    required this.durationText,
    required this.status,
  });

  factory DistanceMatrixResult.fromJson(Map<String, dynamic> json) {
    return DistanceMatrixResult(
      distance: json['distance'] ?? 0,
      distanceText: json['distanceText'] ?? '',
      duration: json['duration'] ?? 0,
      durationText: json['durationText'] ?? '',
      status: json['status'] ?? 'OK',
    );
  }

  int get durationMinutes => (duration / 60).round();
}

class RoutesService {
  final ApiService _apiService = ApiService();

  /// Calcula a rota entre dois pontos
  Future<RouteInfo?> calculateRoute({
    required Waypoint origin,
    required Waypoint destination,
    String travelMode = 'driving',
  }) async {
    try {
      final response = await _apiService.post('/routes/calculate', body: {
        'origin': origin.toJson(),
        'destination': destination.toJson(),
        'travelMode': travelMode,
      });
      return RouteInfo.fromJson(response);
    } catch (e) {
      print('Error calculating route: $e');
      return null;
    }
  }

  /// Obtém a matriz de distâncias entre múltiplas origens e destinos
  Future<List<DistanceMatrixResult>> getDistanceMatrix({
    required List<Waypoint> origins,
    required List<Waypoint> destinations,
    String travelMode = 'driving',
  }) async {
    try {
      final response = await _apiService.post('/routes/distance-matrix', body: {
        'origins': origins.map((w) => w.toJson()).toList(),
        'destinations': destinations.map((w) => w.toJson()).toList(),
        'travelMode': travelMode,
      });

      if (response is List) {
        return response.map((json) => DistanceMatrixResult.fromJson(json)).toList();
      }
      return [];
    } catch (e) {
      print('Error getting distance matrix: $e');
      return [];
    }
  }

  /// Obtém o tempo de viagem de uma origem para um destino
  Future<TravelInfo?> getTravelTime({
    required Waypoint origin,
    required Waypoint destination,
    String travelMode = 'driving',
  }) async {
    try {
      final route = await calculateRoute(
        origin: origin,
        destination: destination,
        travelMode: travelMode,
      );

      if (route != null) {
        return TravelInfo(
          durationMinutes: (route.duration / 60).round(),
          durationText: route.durationText,
          distanceMeters: route.distance,
          distanceText: route.distanceText,
          travelMode: route.travelMode,
        );
      }
      return null;
    } catch (e) {
      print('Error getting travel time: $e');
      return null;
    }
  }

  /// Encontra o melhor modo de transporte (o mais rápido)
  Future<TravelInfo?> getBestTravelMode({
    required Waypoint origin,
    required Waypoint destination,
  }) async {
    final modes = ['walking', 'driving', 'transit'];
    TravelInfo? bestOption;

    for (final mode in modes) {
      try {
        final info = await getTravelTime(
          origin: origin,
          destination: destination,
          travelMode: mode,
        );

        if (info != null) {
          if (bestOption == null || info.durationMinutes < bestOption.durationMinutes) {
            bestOption = TravelInfo(
              durationMinutes: info.durationMinutes,
              durationText: info.durationText,
              distanceMeters: info.distanceMeters,
              distanceText: info.distanceText,
              travelMode: mode,
            );
          }
        }
      } catch (e) {
        // Continuar com o próximo modo
      }
    }

    return bestOption;
  }
}
